### PR TITLE
Make ReactElement a proper Monoid and other ergonomic tweaks

### PR DESCRIPTION
**Pull request by fsoikin**
* By suggestion from @MonoidMusician, `ReactElement` is now a full `Monoid`.
* Fixed a bug with `readForeign` and nested nullables where on error the expected type was reported incorrectly.
* A bunch of small tweaks and fixes to some internals and docs.
* See release notes [in CHANGELOG](https://github.com/collegevine/purescript-elmish/pull/84/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR8-R20)

### CODE REVIEW

File: src/Elmish/React.purs [hunk 1]
```diff
@@ -112,14 +129,6 @@ setState :: ∀ state. ReactComponentInstance -> state -> (Effect Unit) -> Effec
 setState = runEffectFn3 setState_
 foreign import setState_ :: ∀ state. EffectFn3 ReactComponentInstance state (Effect Unit) Unit
 
-getField :: ∀ @a. CanReceiveFromJavaScript a => String -> ReactComponentInstance -> Effect (Maybe a)
-getField field object = runEffectFn2 getField_ field object <#> readForeign @a
-foreign import getField_ :: EffectFn2 String ReactComponentInstance Foreign
-
-setField :: ∀ @a. CanPassToJavaScript a => String -> a -> ReactComponentInstance -> Effect Unit
-setField = runEffectFn3 setField_
-foreign import setField_ :: ∀ a. EffectFn3 String a ReactComponentInstance Unit
```
Comment (fsoikin) [inline src/Elmish/React.purs:121]: This moved to a separate module `Elmish.React.Internal`

File: src/Elmish/Component.purs [hunk 2]
```diff
@@ -307,13 +307,13 @@ bindComponent cmpt stateStrategy = \def -> -- Explicit lambda to make sure `def`
...
-        getSubscriptions = getField @(Array (Effect Unit)) subscriptionsField >>> map (fromMaybe [])
-        setSubscriptions = setField @(Array (Effect Unit)) subscriptionsField
+        subscriptionsField = Field @"__subscriptions" @(Array (Effect Unit))
+        getSubscriptions = getField subscriptionsField >>> map (fromMaybe [])
+        setSubscriptions = setField subscriptionsField
 
-        unmountedField = "__unmounted"
-        getUnmounted = getField @Boolean unmountedField >>> map (fromMaybe false)
-        setUnmounted = setField @Boolean unmountedField
+        unmountedField = Field @"__unmounted" @Boolean
```
Comment (fsoikin) [inline src/Elmish/Component.purs:314]: The `Field` instance binds together field name and type, making mistakes less likely.

**Review by gasi — APPROVED** [review]
LGTM — nice 👍

**Review by mcordova47 — APPROVED** [review]
Sweet!

Now `Utils.HTML.join` can just be replace with `Data.Foldable.intercalate` 👍🏻

**Comment by fsoikin** [comment]
Yep. And `maybeHtml'` is just `folkd`

**Pull request merged**